subsys/usb/class/cdc_acm.c

```diff
@@ -416,29 +418,38 @@ static void cdc_acm_do_cb(struct cdc_acm_dev_data_t *dev_data,
 		LOG_DBG("Device connected");
 		break;
 	case USB_DC_CONFIGURED:
-		cdc_acm_read_cb(cfg->endpoint[ACM_OUT_EP_IDX].ep_addr, 0,
-				dev_data);
+		LOG_INF("Device configured");
+		if (!dev_data->configured) {
+			LOG_ERR("create transfer on ep 0x%02x",
+				cfg->endpoint[ACM_OUT_EP_IDX].ep_addr);
+			cdc_acm_read_cb(cfg->endpoint[ACM_OUT_EP_IDX].ep_addr, 0,
+					dev_data);
+		}
 		dev_data->tx_ready = true;
 		dev_data->tx_irq_ena = true;
 		dev_data->rx_irq_ena = true;
 		dev_data->usb_status = status;
-		LOG_INF("Device configured");
+		dev_data->configured = true;
 		break;
 	case USB_DC_DISCONNECTED:
 		LOG_INF("Device disconnected");
 		cdc_acm_reset_port(dev_data);
 		dev_data->usb_status = status;
+		dev_data->configured = false;
 		break;
 	case USB_DC_SUSPEND:
 		LOG_INF("Device suspended");
 		dev_data->usb_status = status;
 		break;
 	case USB_DC_RESUME:
+		LOG_INF("Device resumed");
 		if (dev_data->usb_status == USB_DC_SUSPEND) {
-			cdc_acm_read_cb(cfg->endpoint[ACM_OUT_EP_IDX].ep_addr,
-					0, dev_data);
-			dev_data->usb_status = USB_DC_CONFIGURED;
-			LOG_INF("Define resumed");
+			if (dev_data->configured) {
+				LOG_INF("from suspend");
+				cdc_acm_read_cb(cfg->endpoint[ACM_OUT_EP_IDX].ep_addr,
+						0, dev_data);
+				dev_data->usb_status = USB_DC_CONFIGURED;
```

Comment:
Instead of using such cheating here, it would be better to use the introduced `configured` flag in `cdc_acm_fifo_fill()` to check if the device is ready to send. I mean to replace:
https://github.com/zephyrproject-rtos/zephyr/blob/c0d7b39c7f402f044530fe25000d4dee6f9aae98/subsys/usb/class/cdc_acm.c#L568
And actually, after such change, the `usb_status` field could be removed entirely - there is no longer a need to check above if `usb_status` is `USB_DC_SUSPEND`, as the transfer is started only when the `configured` flag is set.

Reply:
I'm not sure I got You right. If We change it like this:
``` diff
-	if (dev_data->usb_status != USB_DC_CONFIGURED) {
+	if (dev_data->configured) {
		LOG_WRN("Device not configured, drop %d bytes", len);
		return 0;
	}
```
And the HOST will do:
Suspend->Resume->Configured->Suspend
The cdc will try to send even  `SUSPEND` was received last (we should not send then).
`cdc_acm_fifo_fill` must schedule work only when the last state was `USB_DC_CONFIGURED`

Reply:
Right, my mistake. The `configured` flag alone is not sufficient.
But my point was actually to make it clearer when the device is allowed to send. The current solution with stored `status` is IMO artificial. Especially when you need to change the `usb_staus` field to `USB_DC_CONFIGURED` after getting the callback with `USB_DC_RESUME` parameter. I think it would be better to use a flag like `can_send` instead. Alternatively, a `suspended` flag could be used (and then data could be sent `if (dev_data->configured && !dev_data->suspended)`).

Reply:
Updated.